### PR TITLE
Adding Tests for query endpoint 

### PR DESCRIPTION
**Pull request by sam-i-am012**
Addresses #21 
Also added logging using Log4j 2 (#46)

### CODE REVIEW

**Comment by korydraughn** [comment]
Moving forward, let's try to keep the number of issues addressed in a PR down to one. That makes reviewing the PR much easier/faster.

Focus on resolving a single issue. If you can address multiple issues without them interleaving, that's where opening multiple PRs makes sense.

What's the status of this PR?

**Comment by sam-i-am012** [comment]
Oh sorry, I was only adding it to the query endpoint, not to all the other ones. I just figured it might be convenient to begin adding loggers here and then add them to the rest of the tests as I go along instead of having to go through every test again at the end to add the logging. 
I believe this PR is ready for review, I added an extra commit last night when I realized I forgot to make a variable `private`.

**Comment by korydraughn** [comment]
No problem. Please squash things to taste and/or attach issue numbers to the commits without them and then I'll do a pass over the code.

**Comment by korydraughn** [comment]
We don't receive notifications for force-pushes. Anytime you do that, post a comment so we're notified.

File: src/test/java/org/example/Operations/QueryOperationsTest.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            rods.queryOperations().removeSpecificQuery(rodsToken, specificQueryName);
+        }
+    }
+
+    private int getIrodsResponseStatusCode(String jsonString) {
+        JsonNode rootNode = assertDoesNotThrow(() -> mapper.readTree(jsonString),
+                "JsonProcessingException was thrown");
+        JsonNode statusCodeNode = rootNode.path("irods_response").path("status_code");
+        return statusCodeNode.asInt();
+    }
```
Comment (korydraughn) [inline src/test/java/org/example/Operations/QueryOperationsTest.java:139]: Is this defined in other test modules too? It looks familiar.

**Review by korydraughn — COMMENTED** [review]
Looking good so far.

File: src/test/java/org/example/Operations/QueryOperationsTest.java (hunk 1 above)
Comment (sam-i-am012) [reply src/test/java/org/example/Operations/QueryOperationsTest.java:139]: Yes it's in the other tests as well.

File: src/test/java/org/example/Operations/QueryOperationsTest.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        // Create a client with rodsuser status
+        alice = new Wrapper(baseUrl, "alice", "alicepass");
+        alice.authenticate();
+        aliceToken = alice.getAuthToken();
+    }
+
+    @Test
+    public void testSupportForGenquery1() {
+        String query = "select COLL_NAME";
+        QueryExecuteGenqueryParams params = new QueryExecuteGenqueryParams();
```
[the remark this answers is not in the document]
Comment (sam-i-am012) [reply src/test/java/org/example/Operations/QueryOperationsTest.java:56]: Created #52

File: src/test/java/org/example/Operations/QueryOperationsTest.java (hunk 1 above)
Comment (korydraughn) [reply src/test/java/org/example/Operations/QueryOperationsTest.java:139]: See https://github.com/irods/irods_client_http_java/pull/50#discussion_r1700464128.

File: src/test/java/org/example/Operations/QueryOperationsTest.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        rodsToken = rods.getAuthToken();
+
+        // Create a client with rodsuser status
+        alice = new Wrapper(baseUrl, "alice", "alicepass");
+        alice.authenticate();
+        aliceToken = alice.getAuthToken();
+    }
+
+    @Test
+    public void testSupportForGenquery1() {
```
[the remark this answers is not in the document]
Comment (sam-i-am012) [reply src/test/java/org/example/Operations/QueryOperationsTest.java:54]: Done.

File: src/main/java/org/example/Operations/QueryOperations.java [hunk 4]
```diff
@@ -52,10 +56,12 @@ public Response executeSpecificQuery(String token, String name, QueryExecuteSpec
         Map<Object, Object> formData = new HashMap<>();
         formData.put("op", "execute_specific_query");
         formData.put("name", name);
-        params.getArgs().ifPresent(val -> formData.put("args", val));
-        params.getArgsDelimiter().ifPresent(val -> formData.put("args-delimiter", val));
-        params.getOffset().ifPresent(val -> formData.put("offset", String.valueOf(val)));
-        params.getCount().ifPresent(val -> formData.put("count", String.valueOf(val)));
+        if (params != null) {
```
[the remark this answers is not in the document]
Comment (sam-i-am012) [reply src/main/java/org/example/Operations/QueryOperations.java:59]: Ahh makes sense, I'll make an issue to implement that throughout the other operations as well
Created #53

File: src/main/resources/log4j2.xml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        <File name="LogFile" fileName="app.log">
+            <PatternLayout>
+                <Pattern>%d{yyyy-MM-dd HH:mm:ss} %p %c{1.} [%t] %m%n</Pattern>
+            </PatternLayout>
+        </File>
+    </Appenders>
+    <Loggers>
+        <Root level="debug">
+            <AppenderRef ref="Console" />
+            <AppenderRef ref="LogFile" />
```
[the remark this answers is not in the document]
Comment (sam-i-am012) [reply src/main/resources/log4j2.xml:19]: Done.

File: src/main/java/org/example/Util/HttpRequestUtil.java [hunk 6]
```diff
@@ -75,14 +73,12 @@ public static HttpResponse<String>  sendAndParseGET(Map<Object, Object> formData
                 .build();
 
         // sending request
-        HttpResponse<String> response = null;
+        HttpResponse<String> response;
         try {
             response = client.send(request, HttpResponse.BodyHandlers.ofString());
         } catch (IOException | InterruptedException e) {
             throw new RuntimeException(e);
```
[the remark this answers is not in the document]
Comment (sam-i-am012) [reply src/main/java/org/example/Util/HttpRequestUtil.java:80]: No it doesn't. Would it be better to add the exception to the method signature?

File: src/main/java/org/example/Util/HttpRequestUtil.java [hunk 7]
```diff
@@ -75,14 +73,12 @@ public static HttpResponse<String>  sendAndParseGET(Map<Object, Object> formData
...
         }
-
-
         return response;
```
[the remark this answers is not in the document]
Comment (sam-i-am012) [reply src/main/java/org/example/Util/HttpRequestUtil.java:82]: Yes, I could do this instead:
```java
    try {
        return client.send(request, HttpResponse.BodyHandlers.ofString());
    } catch (IOException | InterruptedException e) {
        throw new RuntimeException(e);
    }
}
```

File: src/main/java/org/example/Operations/QueryOperations.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply src/main/java/org/example/Operations/QueryOperations.java]: Sounds good. Please handle the remaining operations in a separate PR.

File: src/main/java/org/example/Util/HttpRequestUtil.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply src/main/java/org/example/Util/HttpRequestUtil.java:80]: It really depends.

For now, if this is the pattern you've implemented for the other operations, then leave it as is.

We're limited on time so some things will need to be addressed at a later time.

[the remark this answers is not in the document]
Comment (sam-i-am012) [reply src/main/java/org/example/Util/HttpRequestUtil.java:80]: Ok, sounds good.

File: src/test/java/org/example/Operations/QueryOperationsTest.java (hunk 1 above)
Comment (sam-i-am012) [reply src/test/java/org/example/Operations/QueryOperationsTest.java:139]: Ok, sounds good. Created #54

**Comment by sam-i-am012** [comment]
All existing comments have been resolved. Awaiting further review.

File: src/main/java/org/example/Operations/QueryOperations.java [hunk 8]
```diff
@@ -65,12 +71,18 @@ public Response addSpecificQuery(String token, String name, String sql) {
...
         formData.put("name", name);
-        formData.put("sql", sql);
+
+        // need to URL encode the spaces in the query to prevent illegal characters in the query
+        try {
+            String encodedSql = URLEncoder.encode(sql, StandardCharsets.UTF_8.toString());
+            formData.put("sql", encodedSql);
+        } catch (UnsupportedEncodingException e) {
+            e.printStackTrace();
+        }
```
Comment (korydraughn) [inline src/main/java/org/example/Operations/QueryOperations.java:81]: Consider the following:
- The client is running in an environment with no device for stdout/stderr
- `UnsupportedEncodingException` is thrown

What happens to the output of `e.printStackTrace()`?
Will the developer or user have enough information to figure out where things went wrong?

Comment (korydraughn) [reply src/main/java/org/example/Operations/QueryOperations.java:81]: Logging could help, but that doesn't give the developer a chance to react to the exception.

If you let the exception be thrown, the developer can catch it and make decisions based on their own requirements.

Comment (sam-i-am012) [reply src/main/java/org/example/Operations/QueryOperations.java]: Done.

**Comment by sam-i-am012** [comment]
All existing comments have been resolved. Awaiting further review.

File: src/test/java/org/example/Operations/QueryOperationsTest.java [hunk 9]
```diff
@@ -82,7 +84,7 @@ public void testSupportForSpecificQueries() {
     }
 
     @Test
-    public void testAddingAndRemovingSpecificQueries() {
+    public void testAddingAndRemovingSpecificQueries() throws UnsupportedEncodingException {
```
Comment (korydraughn) [inline src/test/java/org/example/Operations/QueryOperationsTest.java:87]: For the test, you can just assertNoThrow on the call. No need for the exception to bubble up to this level.

Comment (sam-i-am012) [reply src/test/java/org/example/Operations/QueryOperationsTest.java]: Done.

**Comment by sam-i-am012** [comment]
I removed some variables that I realized I wasn't using.

**Comment by korydraughn** [comment]
Please run the tests and if they pass, squash the commits to taste.

**Comment by sam-i-am012** [comment]
Done.

File: pom.xml [hunk 10]
```diff
@@ -41,10 +41,14 @@
...
-            <version>2.0.2-beta</version>
-            <scope>test</scope>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-api</artifactId>
+            <version>2.17.0</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-core</artifactId>
+            <version>2.17.0</version>
```
[the remark this answers is not in the document]
Comment (sam-i-am012) [reply pom.xml:51]: Oh no, I didn't catch that. I've updated it.

File: src/main/resources/log4j2.xml [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+        </Console>
+        <File name="LogFile" fileName="app.log">
+            <PatternLayout>
+                <Pattern>%d{yyyy-MM-dd HH:mm:ss} %p %c{1.} [%t] %m%n</Pattern>
+            </PatternLayout>
+        </File>
+    </Appenders>
+    <Loggers>
+        <Root level="debug">
+            <AppenderRef ref="LogFile" />
```
[the remark this answers is not in the document]
Comment (sam-i-am012) [reply src/main/resources/log4j2.xml:18]: Oh yes, I completely misunderstood your previous comment (https://github.com/irods/irods_client_http_java/pull/51#discussion_r1700431677) about logging to a file. 
I've changed it now.

File: src/main/resources/log4j2.xml [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+    <Appenders>
+        <Console name="Console" target="SYSTEM_OUT">
+            <LevelRangeFilter minLevel="ERROR" maxLevel="ERROR" onMatch="ACCEPT" onMismatch="DENY" />
+            <PatternLayout pattern="${layout}"/>
+        </Console>
+        <File name="LogFile" fileName="app.log">
+            <PatternLayout>
+                <Pattern>%d{yyyy-MM-dd HH:mm:ss} %p %c{1.} [%t] %m%n</Pattern>
+            </PatternLayout>
+        </File>
```
[the remark this answers is not in the document]
Comment (sam-i-am012) [reply src/main/resources/log4j2.xml]: Yep, I've removed it now.

**Comment by sam-i-am012** [comment]
I have resolved the comments.

**Review by korydraughn — COMMENTED** [review]
Looks good.

Do the tests pass?

If you change the log level in the log4j config file, do the tests  begin writing messages to the console?

**Comment by sam-i-am012** [comment]
Yes, tests are all passing and if I remove the LevelRangeFilter from only displaying logs at the error level, the messages are written to console.

**Comment by korydraughn** [comment]
I'm not familiar with the level range filter. What does that do?

Normally, the person just changes the log level to get more output (e.g. from `info` to `debug` or `trace`).

**Comment by sam-i-am012** [comment]
It should only write the message to the console when its on the error level, but I got that from watching a youtube tutorial and that's what he did. I could just remove it too if that is not the norm.

**Comment by korydraughn** [comment]
No, it's fine as is.

So for now, if the tests are passing, squash to taste.

**Comment by korydraughn** [comment]
Are you happy with the commits?
Has everything been squashed to your liking?

**Comment by sam-i-am012** [comment]
Sorry, didn't see this comment. I'm squashing now.

**Comment by sam-i-am012** [comment]
Done.

**Review by korydraughn — APPROVED** [review]
Pound it.

For the commit message, `[21, 46] ...`, make sure there's no space following the comma.

**Comment by sam-i-am012** [comment]
Done.

**Comment by korydraughn** [comment]
Merging.

**Pull request merged**